### PR TITLE
[Core][nightly-test] better way of calculating num features

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?

previously, we use num_columns - 2 to calculate the feature size where we assumes both the "label" and arrow internal column `__index_level_0__` exists in the schema.

however, the `__index_level_0__` might not exist in the dataset's schema, which causes this check fail. 
in this pr, we use a better way to filter out arrow internal column and label column.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests https://console.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_xhpjT94JnGDwf9PnUjw78DJe
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/nightly_tests/dataset/ray_sgd_training.py [hunk 1]
```diff
@@ -529,9 +531,13 @@ def train(self):
     preprocessing_end_time = timeit.default_timer()
     print("Preprocessing time (s): ", preprocessing_end_time - e2e_start_time)
 
-    num_columns = len(train_dataset.schema().names)
-    # remove label column and internal Arrow column.
-    num_features = num_columns - 2
+    # filter label column and internal Arrow column.
```
Comment (matthewdeng) [inline release/nightly_tests/dataset/ray_sgd_training.py:534]: Explicitly mention the column name here? `__index_level_0__`

Also in the PR description it says `index_level_0`.

File: release/nightly_tests/dataset/ray_sgd_training.py [hunk 2]
```diff
@@ -529,9 +531,13 @@ def train(self):
...
+    def is_feature_column(column_name):
+        return column_name != "label" and not column_name.startswith("__")
+
+    num_features = len(
+        list(filter(lambda name: is_feature_column(name), train_dataset.schema().names))
```
Comment (matthewdeng) [inline release/nightly_tests/dataset/ray_sgd_training.py:539]: nit: lambda is unnecessary 
```suggestion
        list(filter(is_feature_column, train_dataset.schema().names))
```

**Review by matthewdeng — APPROVED** [review]

**Review by rkooo567 — APPROVED** [review]
